### DISCUSSION TITLE
Direction for CEKit

### DISCUSSION THREAD

**Discussion by greenatatlassian** [cekit/cekit#798] [General]
My team is considering adopting CEKit. We've been exploring using CEKit to build what we often call "base" docker images—images designed to drop in a service or application prior to deployment. My team really likes some of the concepts of CEKit, such as modules, and the out of the box testing option.   

We would like to understand the status and direction of CEKit moving forwards. During our exploration of CEKit, we've identified a few things that we may wish to add/improve/change—and already raised a few PRs reflecting some of what we've found. Thanks for reviewing and helping us get those merged! 

I'm starting this conversation to help us understand how we might best be able to contribute back to the project going forward. I've read through the contribution guidelines already, but we're still curious to know:
- is there a road map for new features, or is this purely in maintenance mode? 
- similarly, is there a process for proposing new features?
- what use cases do other users have? (we've only been able to find the example of Quarkus using CEKit, which has been informative, and see the RedHat integrations in the code base) Seeing some other use cases may help us understand whether a contribution we could make would fit with the projects existing user base.

In our team, we're probably only going to be using CEKit with docker, and expect to lean heavily into the concept of modules (hence #783 and #792). We also may want to write another test framework integration for JUnit5.

Thanks in advance!

**Comment by rnc** [comment]
Hi!

We're glad to hear you're finding CEKit useful!

CEKit is definitely not in 'maintenance mode' although there isn't a strict roadmap as such. Features such as #787 tend to be added on-demand by request of the various product teams. We have also recently for instance made changes to improve the code base for long term maintenance.

GH Issues / Discussions would be the best place to start talking about potential new features. As CEKit is quite heavily used by various RH teams (as you can see below) any new additions would have to not break the current feature set or be coded in such a way that they are optional. 

Groups currently using CEKit include:

* https://github.com/quarkusio/quarkus-images/
* https://github.com/wildfly/wildfly-s2i/
* https://github.com/jboss-container-images/jboss-eap-8-openshift-image
* https://github.com/jboss-container-images/jboss-eap-7-openshift-image
* https://github.com/jboss-container-images/rhpam-7-openshift-image
* https://github.com/kiegroup/kogito-images
* https://github.com/jboss-container-images/jboss-webserver-5-openshift-image
* https://github.com/jboss-container-images/jboss-webserver-5-image
* https://github.com/jboss-container-images/jboss-webserver-3-image
* https://github.com/jboss-container-images/jboss-webserver-3-openshift-image
* https://github.com/web-servers/jboss-jws-container-images-modules
* https://github.com/jboss-container-images/openjdk

As a slight aside, I've just enabled GH Discussions if that would be a more appropriate location for these sorts of discussions (I can always move this if you want)

**Comment by jmtd** [comment]
Hi @greenatatlassian 

I used to work for a team within Red Hat named "Cloud Enablement" who had responsibility for images of lots of Red Hat JBoss products, and the team developed cekit (and some other stuff) as part of figuring out *how* we would do container images.

A few years back the CE team was disbanded ("mission complete") and we moved to other teams. I moved to Red Hat's OpenJDK team and look after the Red Hat OpenJDK container images (a family of 12 images now, in the [jboss-container-images/openjdk](https://github.com/jboss-container-images/openjdk) repo that @rnc linked, above)

We (still) use cekit for these images. I still contribute to cekit, although much less than I used to, but I am trying to pick that up a bit. (I've been trying to help @rnc review your recent, and very welcome, contributions. Thank you!)

The prospect of other folks such as yourself using it is very appealing. I made a hard choice to continue to use cekit when I moved to the OpenJDK team: at the time, we had just one image, and the cost/complexity balance was less clear for a tool which hadn't gained adoption outside of Red Hat at the time. I think the choice paid off as the number of similar images we look after exploded.

> is there a road map for new features, or is this purely in maintenance mode?
> similarly, is there a process for proposing new features?

I don't have a good answer for these two questions myself, but, I think we'd happily work out a project governance that was attractive to you, as having other users investing in cekit is very attractive to us. What kind of assurances do you need?

> We also may want to write another test framework integration for JUnit5.

I'd be very interested to learn more about your thoughts here. The behave testing aspect of cekit is something I value, although it's a bit rough in places, and one awkward detail for us is it depends on docker, whereas many of my colleagues have moved the rest of their workloads to other container engines. (I note that you would be using the docker backend if you adopted cekit. There's no risk of it going away, at any rate)

I guess it would not be of use to you, but I was considering investing time in expanding the behave test framework to better support docker alternatives (that work would likely all land in https://github.com/cekit/behave-test-steps)

**Comment by jmtd** [comment]
Oh I should probably add (but didn't want to speak for him that @rnc is part of a team who actively work on Cekit.

**Comment by greenatatlassian** [comment]
Thanks @jmtd and @rnc. Those insights are very helpful. I've looped in my team. We'll discuss and see how those directions align with our own goals. 

I'm taking a step back to work with another team for a while, so will be a little less involved, but still around.
